File: python/simulate.py
import numpy as np 
import random
import pandas as pd

# Set resistance, inertance and elastance 
print("Input desired values for R, I and I. "
      "Type a value and press enter.")
R = float(input("R = "))
I = float(input("I = "))
E = float(input("E = "))

# Compute resonance frequency for reference
fr = np.sqrt(E/I)/(2*np.pi)
print("The resonance frequency for these parameters is: fr = %.2fHz" % (fr))

# Set the maximum and minimum frequencies
print("\nThe pressure and flow data will be "
      "computed at specific equally\n"
      "spaced frequencies. Choose the smallest " 
      "and largest frequencies,\n"
      "and the number of points.")

while True:
    small_freq = float(input("Smallest Frequency (Hz) = "))
    large_freq = float(input("Largest Frequency (Hz) = "))
    num_points = int(input("Number of points = "))
    
    # Sanity check the input values
    if small_freq >= large_freq:
        print("\nThe smallest frequency must be strictly smaller\n"
              "than the largest frequency. Try again.\n")
    elif num_points <= 0:
        print("\nThe number of points must be strictly positive.\n"
              "Try again.\n")
    else:
        # Success, move on to the next step
        break
    

# Generate the frequency vector
f = np.linspace(small_freq, large_freq, num_points) # Frequency in Hertz
omega = 2*np.pi*f # Convert to radial frequency (rads/second)

# Compute the impedance
print("\nThe impedance Z(f) will be computed at", num_points ,
      "frequency point(s):")

# Print a selection of frequencies
if len(f) < 10:
    for n in range(len(f)-1):
        print("%.2fHz, " % (f[n]), end="")
    print("%.2fHz" % (f[-1]))
else:
    print("%.2fHz, %.2fHz, %.2fHz, ..., %.2fHz, %.2fHz."
          % (f[0], f[1], f[2], f[-2], f[-1]))

# Compute the impedance
Z = R + 1j*omega*I + E/(1j*omega)

# Compute random phases for the pressure signal
pressure_phase = np.random.uniform(low=0.0, high=2*np.pi, size=len(f))

# Assuming that the magnitude of the pressure signal is
# one, the magnitude of the flow signal in the frequency
# domain is given by the magnitude of the impedance Z(f)
flow_mag = abs(Z)

# To obtain the phase of the flow in the frequency domain,
# subtract the argument of the impedance from the phase of
# the pressure signal
flow_phase = pressure_phase - np.angle(Z)

# To generate the time domain version of the pressure and
# flow signals, it is necessary to know the sampling
# frequency and the desired number of samples
num_samples = int(input("\nInput the desired total number of "
                        "samples in the simulated data: "))
sampling_freq = float(input("Input the desired sampling frequency (Hz): "))

# The time domain pressure and flow signals are now constructed
# by summing sinusoidal components with the correct magnitude
# and phase (the Fourier series of the signal), and evaluating it
# and the correct sampling points.

data = {"time": [], "pressure": [], "flow": []}

for n in range(num_samples):

    # Store current timestamp
    t = n/sampling_freq
    data["time"].append(t)
    
    # Construct the pressure signal
    data["pressure"].append(np.cos(omega*t - pressure_phase).sum())
    
    # Construct the flow signal
    data["flow"].append((flow_mag * np.cos(omega*t - flow_phase)).sum())


# Make a pandas dataframe so that writing the csv file is easier
df = pd.DataFrame(data)

# Get a filename to store the simulated data
print("Choose a filename to store the simulated data\n(.csv will be appended)\n")
filename = input("Filename: ") + ".csv"
df.to_csv(filename, index=False)
print("Written",filename)

File: octave/simulate.m
# Set resistance, inertance and elastance 
fprintf("Input desired values for R, L and C. \
Type a value and press enter.\n")

R = input("R = ");
L = input("L = ");
C = input("C = ");

# Compute resonance frequency for reference
fr = 1 / (2 * pi * sqrt(L*C));
fprintf("The resonance frequency for these parameters is: \
fr = %.2fHz\n", fr)

## Set the maximum and minimum frequencies
fprintf("\nThe pressure and flow data will be \
computed at specific equally\n\
spaced frequencies. Choose the smallest \
and largest frequencies,\nand the number of points.\n")

while true
  small_freq = input("Smallest Frequency (Hz) = ");
  large_freq = input("Largest Frequency (Hz) = ");
  num_points = input("Number of points = ");

  ## Sanity check the input values
  if small_freq >= large_freq
    fprintf("\nThe smallest frequency must be strictly smaller\n than the largest frequency. Try again.\n")
  elseif num_points <= 0
    fprintf("\nThe number of points must be strictly positive.\nTry again.\n")
  else
    ## Success, move on to the next step
    break
  end
end

## Generate the frequency vector
f = linspace(small_freq, large_freq, num_points); # Frequency in Hertz
omega = 2*pi*f; # Convert to radial frequency (rads/second)

## Compute the impedance
fprintf("\nThe impedance Z(f) will be computed at %d frequency point(s):\n", num_points)

## Print a selection of frequencies
if length(f) < 10
  for n = 1:length(f)-1
    fprintf("%.2fHz, ", f(n))
  end
  fprintf("%.2fHz\n", f(end))
else
  fprintf("%.2fHz, %.2fHz, %.2fHz, ..., %.2fHz, %.2fHz.\n",
	  f(1), f(2), f(3), f(end-1), f(end))
end

## Compute the impedance
Z = R + 1j*omega*L + 1./(1j*omega*C);

## Compute random phases for the pressure signal
pressure_phase = 2 * pi * rand(1,length(f));

# Assuming that the magnitude of the pressure signal is
# one, the magnitude of the flow signal in the frequency
# domain is given by the magnitude of the impedance Z(f)
flow_mag = abs(Z);

# To obtain the phase of the flow in the frequency domain,
# subtract the argument of the impedance from the phase of
# the pressure signal
flow_phase = pressure_phase - angle(Z);

# To generate the time domain version of the pressure and
# flow signals, it is necessary to know the sampling
# frequency and the desired number of samples
num_samples = input("\nInput the desired total number of samples in the simulated data: ");
sampling_freq = input("Input the desired sampling frequency (Hz): ");

pressure_mag = ones(1, length(f))
data = [f' Z' pressure_mag' pressure_phase' flow_mag' flow_phase']

# Get a filename to store the simulated data and write the csv file
fprintf("Choose a filename to store the frequency domain data\n(.csv will be appended)\n")
filebase = input("Filename: ","s");
filename = strcat(filebase, ".csv");
fid = fopen(filename, 'w+');
fprintf(fid,"freq,Z,pressure_mag,pressure_phase,flow_mag,flow_phase\n");
fclose(fid);
csvwrite(filename,data,"-append");
fprintf("Written %s\n",filename);

# The time domain pressure and flow signals are now constructed
# by summing sinusoidal components with the correct magnitude
# and phase (the Fourier series of the signal), and evaluating it
# and the correct sampling points.

##for n in range(num_samples):
time = [];
pressure = [];
flow = [];
for n = 1:num_samples
  ## Store current timestamp
  t = (n-1)/sampling_freq;
  time = [time; t]; # Append to time vector
  
  ## Construct the pressure signal at time t
  p = sum(cos(omega*t - pressure_phase)); # Sum over Fourier components
  pressure = [pressure; p];
  
  ## Construct the flow signal at time t
  f = sum(flow_mag .* cos(omega*t - flow_phase));
  flow = [flow; f];
end

time
pressure
flow

data = [time pressure flow]

## Get a filename to store the simulated data and write the csv file
fprintf("Choose a filename to store the time domain data\n(.csv will be appended)\n")
filebase = input("Filename: ","s");
filename = strcat(filebase, ".csv");
fid = fopen(filename, 'w+');
fprintf(fid,"time,pressure,flow\n");
fclose(fid);
csvwrite(filename,data,"-append");
fprintf("Written %s\n",filename);
